### PR TITLE
Support for Kubernetes v1.26

### PR DESCRIPTION
**Pull request by ialidzhikov**
**How to categorize this PR?**

/area open-source usability
/kind enhancement
/topology garden seed shoot

**What this PR does / why we need it**:
Support for Kubernetes v1.26

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/6773

**Special notes for your reviewer**:
- I didn't vendor new `k8s.io/*` Golang sources, this item is tracked separately in https://github.com/gardener/gardener/issues/7230.
- ⚠️ This PR only contains the Gardener part for supporting 1.26 - there will/must be follow-up PRs for each extension repository where individual support gets added.
- Code changes have been tested in the following scenarios:
  - Create/Delete a new Shoot cluster with 1.26.0
  - Upgrade an existing Shoot cluster to 1.26.0
  - Register a 1.26.0 as a Seed cluster (with HVPA feature gate disabled)

**Release note**:

```feature operator
Gardener can now support Shoot clusters (and Seed clusters with `HVPA` feature gate disabled) with Kubernetes version 1.26. In order to allow creation/update of 1.26 clusters you will have to update the version of your provider extension(s) to a version that supports 1.26 as well. Please consult the respective releases and notes in the provider extension's repository.
```

```feature developer
Gardener can now support Shoot clusters (and Seed clusters with `HVPA` feature gate disabled) with Kubernetes version 1.26. Extension developers have to prepare individual extensions as well to work with 1.26.
```

```user other
The `spec.kubernetes.kubeControllerManager.podEvictionTimeout` field in the `Shoot` API is deprecated in favor of the newly introduced fields `spec.kubernetes.kubeAPIServer.default{NotReady,Unreachable}TolerationSeconds`. The `podEvictionTimeout` field does not have effect when the taint-based eviction is enabled. The taint-based eviction is beta (enabled by default) since Kubernetes 1.13 and GA since Kubernetes 1.18. The newly introduced `defaultNotReadyTolerationSeconds` and `defaultUnreachableTolerationSeconds` fields should be used instead.
```

```user other
The `gardener-apiserver` defaults the `.spec.kubernetes.enableStaticTokenKubeconfig` field to `false` for `Shoot`s with Kubernetes version >= `1.26`. For `Shoot`s with Kubernetes version < `1.26`, the `gardener-apiserver` continues to default the field to `true`. The static token kubeconfig is deprecated. The [`shoots/adminkubeconfig` subresource](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md#shootsadminkubeconfig-subresource) should be used instead.
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,19 +34,19 @@ Please find more information regarding the concepts and a detailed description o
 
 ## K8s Conformance Test Coverage <img src="https://raw.githubusercontent.com/cncf/artwork/master/projects/kubernetes/certified-kubernetes/versionless/color/certified-kubernetes-color.svg" alt="certified kubernetes logo" width="50" align="right"/>
 
-Gardener takes part in the [Certified Kubernetes Conformance Program](https://www.cncf.io/certification/software-conformance/) to attest its compatibility with the K8s conformance testsuite. Currently Gardener is certified for K8s versions up to v1.25, see [the conformance spreadsheet](https://docs.google.com/spreadsheets/d/1uF9BoDzzisHSQemXHIKegMhuythuq_GL3N1mlUUK2h0/edit#gid=0&range=126:127).
+Gardener takes part in the [Certified Kubernetes Conformance Program](https://www.cncf.io/certification/software-conformance/) to attest its compatibility with the K8s conformance testsuite. Currently Gardener is certified for K8s versions up to v1.26, see [the conformance spreadsheet](https://docs.google.com/spreadsheets/d/1uF9BoDzzisHSQemXHIKegMhuythuq_GL3N1mlUUK2h0/edit#gid=0&range=126:127).
```
Comment (shafeeqes) [inline README.md]: Should we be updating this now itself? Gardener does not have `x` in 1.26 column in the spreadsheet currently. I think we can do it when we update the conformance test results, WDYT?

File: test/e2e/gardener/shoot/create_update_delete.go [hunk 2]
```diff
@@ -40,15 +40,15 @@ var _ = Describe("Shoot Tests", Label("Shoot", "default"), func() {
 
 	// explicitly use one version below the latest supported minor version so that Kubernetes version update test can be
 	// performed
-	f.Shoot.Spec.Kubernetes.Version = "1.24.8"
+	f.Shoot.Spec.Kubernetes.Version = "1.25.4"
```
Comment (shafeeqes) [inline test/e2e/gardener/shoot/create_update_delete.go:46]: Nit : `1.25.5`

File: pkg/apis/core/v1alpha1/defaults_test.go [hunk 3]
```diff
@@ -654,6 +637,32 @@ var _ = Describe("Defaults", func() {
 			Expect(obj.Spec.Kubernetes.KubeAPIServer.Logging.HTTPAccessVerbosity).To(BeNil())
 		})
 
+		It("should default the defaultNotReadyTolerationSeconds field", func() {
+			SetDefaults_Shoot(obj)
+			Expect(obj.Spec.Kubernetes.KubeAPIServer.DefaultNotReadyTolerationSeconds).To(PointTo(Equal(int64(300))))
+		})
+
+		It("should not default the defaultNotReadyTolerationSeconds field", func() {
```
Comment (shafeeqes) [inline pkg/apis/core/v1alpha1/defaults_test.go]:
```suggestion
		It("should not overwrite the defaultNotReadyTolerationSeconds field", func() {
```
or 
```suggestion
		It("should not default the defaultNotReadyTolerationSeconds field if it is already set", func() {
```

File: pkg/apis/core/v1alpha1/defaults_test.go [hunk 4]
```diff
@@ -654,6 +637,32 @@ var _ = Describe("Defaults", func() {
...
+			SetDefaults_Shoot(obj)
+			Expect(obj.Spec.Kubernetes.KubeAPIServer.DefaultNotReadyTolerationSeconds).To(PointTo(Equal(tolerationSeconds)))
+		})
+
+		It("should default the defaultUnreachableTolerationSeconds field", func() {
+			SetDefaults_Shoot(obj)
+			Expect(obj.Spec.Kubernetes.KubeAPIServer.DefaultUnreachableTolerationSeconds).To(PointTo(Equal(int64(300))))
+		})
+
+		It("should not default the defaultUnreachableTolerationSeconds field", func() {
```
Comment (shafeeqes) [inline pkg/apis/core/v1alpha1/defaults_test.go]: same as above

File: pkg/apis/core/v1beta1/defaults_test.go [hunk 5]
```diff
@@ -782,20 +805,6 @@ var _ = Describe("Defaults", func() {
...
-			})
-
-			Context("allowPrivilegedContainers field is not set", func() {
-				It("should not set the field", func() {
-					SetDefaults_Shoot(obj)
-
-					Expect(obj.Spec.Kubernetes.AllowPrivilegedContainers).To(BeNil())
-				})
-			})
-		})
```
Comment (shafeeqes) [inline pkg/apis/core/v1beta1/defaults_test.go]: Why did you remove this?

**Review by shafeeqes — COMMENTED** [review]
Thanks for the well structured PR.
Had a combined look with @ary1992 and left some comments and questions.

File: docs/usage/shoot_access.md [hunk 6]
```diff
@@ -21,6 +21,12 @@ It is **not** the recommended method to access the shoot cluster as the static t
 - The static token in the `kubeconfig` doesn't have any expiration date. Read [this document](shoot_credentials_rotation.md#kubeconfig) to learn how to rotate the static token.
 - The static token doesn't have any user identity associated with it. The user in that token will always be `system:cluster-admin` irrespective of the person accessing the cluster. Hence, it is impossible to audit the events in cluster.
 
+When `enableStaticTokenKubeconfig` field is not explicitly set in the Shoot spec:
+- for Shoot clusters using Kubernetes version < 1.26 the field is defaulted to `true`.
+- for Shoot clusters using Kubernetes version >= 1.26 the field is defaulted to `false`.
+
+> Note: Starting Kubernetes 1.27 the `enableStaticTokenKubeconfig` field will be locked to `false`. The [`shoots/adminkubeconfig` subresource](#shootsadminkubeconfig-subresource) should be used instead.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
> Note: Starting with Kubernetes 1.27, the `enableStaticTokenKubeconfig` field will be locked to `false`. The [`shoots/adminkubeconfig` subresource](#shootsadminkubeconfig-subresource) should be used instead.
```

File: pkg/apis/core/v1beta1/defaults_test.go (hunk 5 above)
Comment (ialidzhikov) [reply pkg/apis/core/v1beta1/defaults_test.go]: Thanks for catching this! I believe I wrongly removed it as part of refactoring that was later reverted. I added the removed test again.

**Comment by ialidzhikov** [comment]
`pull-gardener-e2e-kind-upgrade`, `pull-gardener-e2e-kind-ha-single-zone-upgrade` and `pull-gardener-e2e-kind-ha-multi-zone-upgrade` are failing with:
```
  Expected success, but got an error:
      <*errors.StatusError | 0xc0002b0960>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "shoots.core.gardener.cloud \"e2e-upgrade\" is forbidden: [spec.kubernetes.version: Unsupported value: \"1.26.0\": supported values: \"1.25.4\", \"1.24.8\", \"1.23.6\", \"1.22.0\", \"1.21.0\", \"1.20.0\"]",
              Reason: "Forbidden",
              Details: {
                  Name: "e2e-upgrade",
                  Group: "core.gardener.cloud",
                  Kind: "shoots",
                  UID: "",
                  Causes: nil,
                  RetryAfterSeconds: 0,
              },
              Code: 403,
          },
      }
      shoots.core.gardener.cloud "e2e-upgrade" is forbidden: [spec.kubernetes.version: Unsupported value: "1.26.0": supported values: "1.25.4", "1.24.8", "1.23.6", "1.22.0", "1.21.0", "1.20.0"]
  In [It] at: /home/prow/go/src/github.com/gardener/gardener/test/e2e/gardener/upgrade/upgrade.go:71
```

The reason for that seems to be that it first deploys the previous release (1.62.1) where `1.26.0` is not present in the CloudProfile (ref https://github.com/gardener/gardener/blob/v1.62.1/example/provider-local/garden/base/cloudprofile.yaml) and it tries to create a Shoot with 1.26.0.
To fix this I can downgrade to 1.25 in the tests and put a TODO to upgrade to 1.26 when there is a release with this PR. WDYT?

**Comment by rfranzke** [comment]
@ialidzhikov Sounds reasonable! We don't have to wait for the next release, though. You can change https://github.com/gardener/gardener/blob/master/Makefile#L36 after this PR got merged. :)

File: test/e2e/gardener/shoot/create_update_delete.go (hunk 2 above)
Comment (ialidzhikov) [reply test/e2e/gardener/shoot/create_update_delete.go:46]: This comment is not resolved due to https://github.com/gardener/gardener/pull/7275#issuecomment-1381773647.

**Review by shafeeqes — COMMENTED** [review]
/lgtm
